File: DailyChallenges/Scala/LeetCode_1026.scala
object Solution {
    def maxAncestorDiff(root: TreeNode): Int = {
        var ans = -1
        def dfs(node: TreeNode, lmax: Int, lmin: Int): Unit = {
            val _lmax = math.max(lmax, node.value)
            val _lmin = math.min(lmin, node.value)
            if (node.left != null) dfs(node.left, _lmax, _lmin)
            if (node.right != null) dfs(node.right, _lmax, _lmin)
            if (node.left == null && node.right == null) ans = math.max(ans, _lmax - _lmin)    
        }
        dfs(root, root.value, root.value)
        ans
    }
}

File: DailyChallenges/CPP/LeetCode_1026.cpp
// 1026. Maximum Difference Between Node and Ancestor @ 2022/12/09
 
class Solution {
public:

    int maxAncestorDiff(TreeNode* root) {
        int ans = -1;

        function<void(TreeNode*, int, int)> dfs = [&](TreeNode* node, int lmax, int lmin){
            lmax = max(lmax, node->val);
            lmin = min(lmin, node->val);
            if (node->left) dfs(node->left, lmax, lmin);
            if (node->right) dfs(node->right, lmax, lmin);
            if (!node->left & !node->right) ans = max(ans, lmax-lmin);
        };

        dfs(root,root->val, root->val);
        return ans;
    }
};
